Fails tests with OpenSSL master (3.0)

```
$ cmake .. -DCMAKE_C_FLAGS="-I/Users/ur20980/openssl-1.1/include -L/Users/ur20980/openssl-1.1/lib"
-- Found OpenSSL: /opt/local/lib/libcrypto.dylib (found suitable version "1.1.1c", minimum required is "1.1.1")  
-- Looking for clock_gettime
-- Looking for clock_gettime - found
-- Looking for clock_gettime in rt
-- Looking for clock_gettime in rt - not found
-- LITTLE_ENDIAN
-- Found PkgConfig: /opt/local/bin/pkg-config (found version "0.29.2") 
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/ur20980/src/grasshopper-engine/build
$ make -j4 all
Scanning dependencies of target test_tlstree
Scanning dependencies of target gost_core
[  1%] Building C object CMakeFiles/test_tlstree.dir/test_tlstree.c.o
[  3%] Building C object CMakeFiles/gost_core.dir/gosthash2012.c.o
[  7%] Building C object CMakeFiles/gost_core.dir/gost89.c.o
[  7%] Building C object CMakeFiles/gost_core.dir/gosthash.c.o
[  9%] Linking C executable bin/test_tlstree
[ 11%] Building C object CMakeFiles/gost_core.dir/gost_grasshopper_core.c.o
[ 11%] Built target test_tlstree
[ 12%] Building C object CMakeFiles/gost_core.dir/gost_grasshopper_defines.c.o
[ 14%] Building C object CMakeFiles/gost_core.dir/gost_grasshopper_galois_precompiled.c.o
[ 16%] Building C object CMakeFiles/gost_core.dir/gost_grasshopper_precompiled.c.o
[ 18%] Building C object CMakeFiles/gost_core.dir/gost_grasshopper_cipher.c.o
[ 20%] Building C object CMakeFiles/gost_core.dir/gost_ec_keyx.c.o
[ 22%] Building C object CMakeFiles/gost_core.dir/gost_ec_sign.c.o
[ 24%] Building C object CMakeFiles/gost_core.dir/gost_omac.c.o
[ 25%] Building C object CMakeFiles/gost_core.dir/gost_omac_acpkm.c.o
[ 27%] Linking C static library libgost_core.a
[ 27%] Built target gost_core
Scanning dependencies of target gost12sum
Scanning dependencies of target gostsum
Scanning dependencies of target gost_engine
[ 31%] Building C object CMakeFiles/gost12sum.dir/gost12sum.c.o
[ 31%] Building C object CMakeFiles/gostsum.dir/gostsum.c.o
[ 35%] Building C object CMakeFiles/gost_engine.dir/gost_asn1.c.o
[ 35%] Building C object CMakeFiles/gost_engine.dir/e_gost_err.c.o
[ 37%] Building C object CMakeFiles/gost_engine.dir/gost_crypt.c.o
[ 38%] Linking C executable bin/gostsum
[ 40%] Linking C executable bin/gost12sum
[ 42%] Building C object CMakeFiles/gost_engine.dir/gost_ctl.c.o
[ 42%] Built target gostsum
[ 44%] Building C object CMakeFiles/gost_engine.dir/gost_eng.c.o
[ 44%] Built target gost12sum
[ 46%] Building C object CMakeFiles/gost_engine.dir/gost_keywrap.c.o
[ 48%] Building C object CMakeFiles/gost_engine.dir/gost_params.c.o
[ 50%] Building C object CMakeFiles/gost_engine.dir/gost_keyexpimp.c.o
[ 51%] Building C object CMakeFiles/gost_engine.dir/gost_ameth.c.o
/Users/ur20980/src/grasshopper-engine/gost_keyexpimp.c:47:12: error: 'EVP_MD_CTX_ctrl' is
      deprecated [-Werror,-Wdeprecated-declarations]
        || EVP_MD_CTX_ctrl(mac, EVP_MD_CTRL_SET_KEY, 32, mac_key) <= 0
           ^
/Users/ur20980/openssl-1.1/include/openssl/evp.h:553:1: note: 'EVP_MD_CTX_ctrl' has been
      explicitly marked deprecated here
DEPRECATEDIN_3(int EVP_MD_CTX_ctrl(EVP_MD_CTX *ctx, int cmd, int p1, void *p2))
^
/Users/ur20980/openssl-1.1/include/openssl/opensslconf.h:170:35: note: expanded from macro
      'DEPRECATEDIN_3'
#  define DEPRECATEDIN_3(f)       DECLARE_DEPRECATED(f)
                                  ^
/Users/ur20980/openssl-1.1/include/openssl/opensslconf.h:117:56: note: expanded from macro
      'DECLARE_DEPRECATED'
#    define DECLARE_DEPRECATED(f)    f __attribute__ ((deprecated));
                                                       ^
/Users/ur20980/src/grasshopper-engine/gost_keyexpimp.c:48:12: error: 'EVP_MD_CTX_ctrl' is
      deprecated [-Werror,-Wdeprecated-declarations]
        || EVP_MD_CTX_ctrl(mac, EVP_MD_CTRL_MAC_LEN, mac_len, NULL) <= 0
           ^
/Users/ur20980/openssl-1.1/include/openssl/evp.h:553:1: note: 'EVP_MD_CTX_ctrl' has been
      explicitly marked deprecated here
DEPRECATEDIN_3(int EVP_MD_CTX_ctrl(EVP_MD_CTX *ctx, int cmd, int p1, void *p2))
^
/Users/ur20980/openssl-1.1/include/openssl/opensslconf.h:170:35: note: expanded from macro
      'DEPRECATEDIN_3'
#  define DEPRECATEDIN_3(f)       DECLARE_DEPRECATED(f)
                                  ^
/Users/ur20980/openssl-1.1/include/openssl/opensslconf.h:117:56: note: expanded from macro
      'DECLARE_DEPRECATED'
#    define DECLARE_DEPRECATED(f)    f __attribute__ ((deprecated));
                                                       ^
/Users/ur20980/src/grasshopper-engine/gost_keyexpimp.c:140:12: error: 'EVP_MD_CTX_ctrl' is
      deprecated [-Werror,-Wdeprecated-declarations]
        || EVP_MD_CTX_ctrl(mac, EVP_MD_CTRL_SET_KEY, 32, mac_key) <= 0
           ^
/Users/ur20980/openssl-1.1/include/openssl/evp.h:553:1: note: 'EVP_MD_CTX_ctrl' has been
      explicitly marked deprecated here
DEPRECATEDIN_3(int EVP_MD_CTX_ctrl(EVP_MD_CTX *ctx, int cmd, int p1, void *p2))
^
/Users/ur20980/openssl-1.1/include/openssl/opensslconf.h:170:35: note: expanded from macro
      'DEPRECATEDIN_3'
#  define DEPRECATEDIN_3(f)       DECLARE_DEPRECATED(f)
                                  ^
/Users/ur20980/openssl-1.1/include/openssl/opensslconf.h:117:56: note: expanded from macro
      'DECLARE_DEPRECATED'
#    define DECLARE_DEPRECATED(f)    f __attribute__ ((deprecated));
                                                       ^
/Users/ur20980/src/grasshopper-engine/gost_keyexpimp.c:141:12: error: 'EVP_MD_CTX_ctrl' is
      deprecated [-Werror,-Wdeprecated-declarations]
        || EVP_MD_CTX_ctrl(mac, EVP_MD_CTRL_MAC_LEN, mac_len, NULL) <= 0
           ^
/Users/ur20980/openssl-1.1/include/openssl/evp.h:553:1: note: 'EVP_MD_CTX_ctrl' has been
      explicitly marked deprecated here
DEPRECATEDIN_3(int EVP_MD_CTX_ctrl(EVP_MD_CTX *ctx, int cmd, int p1, void *p2))
^
/Users/ur20980/openssl-1.1/include/openssl/opensslconf.h:170:35: note: expanded from macro
      'DEPRECATEDIN_3'
#  define DEPRECATEDIN_3(f)       DECLARE_DEPRECATED(f)
                                  ^
/Users/ur20980/openssl-1.1/include/openssl/opensslconf.h:117:56: note: expanded from macro
      'DECLARE_DEPRECATED'
#    define DECLARE_DEPRECATED(f)    f __attribute__ ((deprecated));
                                                       ^
4 errors generated.
make[2]: *** [CMakeFiles/gost_engine.dir/gost_keyexpimp.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 53%] Building C object CMakeFiles/gost_engine.dir/gost_md.c.o
[ 55%] Building C object CMakeFiles/gost_engine.dir/gost_md2012.c.o
make[1]: *** [CMakeFiles/gost_engine.dir/all] Error 2
make: *** [all] Error 2
$ 
```

Current master fails to compile under `OpenSSL_1_1_1c` as well. 

Branch `openssl_1_1_0_release2` compiles with `OpenSSL_1_1_1c` fine.

After straightening out the arguments via appropriate explicit casts, everything compiles, but some tests fail:
```
Running tests...
/opt/local/bin/ctest --force-new-ctest-process 
Test project /Users/ur20980/src/engine/build
      Start  1: digest
 1/10 Test  #1: digest ...........................   Passed    0.01 sec
      Start  2: curves
 2/10 Test  #2: curves ...........................Child aborted***Exception:   0.01 sec
      Start  3: parameters
 3/10 Test  #3: parameters .......................   Passed    0.03 sec
      Start  4: sign/verify
 4/10 Test  #4: sign/verify ......................***Failed    0.01 sec
      Start  5: TLS
 5/10 Test  #5: TLS ..............................Child aborted***Exception:   0.01 sec
      Start  6: context
 6/10 Test  #6: context ..........................   Passed    0.01 sec
      Start  7: grasshopper
 7/10 Test  #7: grasshopper ......................***Failed    0.04 sec
      Start  8: keyexpimp
 8/10 Test  #8: keyexpimp ........................   Passed    0.01 sec
      Start  9: gost89
 9/10 Test  #9: gost89 ...........................   Passed    0.00 sec
      Start 10: engine
10/10 Test #10: engine ...........................***Failed    1.57 sec

50% tests passed, 5 tests failed out of 10

Total Test time (real) =   1.69 sec

The following tests FAILED:
	  2 - curves (Child aborted)
	  4 - sign/verify (Failed)
	  5 - TLS (Child aborted)
	  7 - grasshopper (Failed)
	 10 - engine (Failed)
Errors while running CTest
make: *** [test] Error 8
```
[CMakeLists.txt](https://github.com/gost-engine/engine/files/3535730/CMakeLists.txt)
[cmake-out.txt](https://github.com/gost-engine/engine/files/3535729/cmake-out.txt)
[make-out.txt](https://github.com/gost-engine/engine/files/3535728/make-out.txt)
[test-out.txt](https://github.com/gost-engine/engine/files/3535727/test-out.txt)
[LastTest.log.txt](https://github.com/gost-engine/engine/files/3535732/LastTest.log.txt)

@beldmit Does the above help?

Also, to my surprise (same compilation, everything same, just different invocation):
```
$ echo $OPENSSL_CONF

$ OPENSSL_CONF="" make test
Running tests...
/opt/local/bin/ctest --force-new-ctest-process 
Test project /Users/ur20980/src/engine/build
      Start  1: digest
 1/10 Test  #1: digest ...........................   Passed    0.00 sec
      Start  2: curves
 2/10 Test  #2: curves ...........................   Passed    0.03 sec
      Start  3: parameters
 3/10 Test  #3: parameters .......................   Passed    0.03 sec
      Start  4: sign/verify
 4/10 Test  #4: sign/verify ......................   Passed    0.04 sec
      Start  5: TLS
 5/10 Test  #5: TLS ..............................   Passed    0.53 sec
      Start  6: context
 6/10 Test  #6: context ..........................   Passed    0.01 sec
      Start  7: grasshopper
 7/10 Test  #7: grasshopper ......................   Passed    0.03 sec
      Start  8: keyexpimp
 8/10 Test  #8: keyexpimp ........................   Passed    0.00 sec
      Start  9: gost89
 9/10 Test  #9: gost89 ...........................   Passed    0.00 sec
      Start 10: engine
10/10 Test #10: engine ...........................***Failed    1.67 sec

90% tests passed, 1 tests failed out of 10

Total Test time (real) =   2.37 sec

The following tests FAILED:
	 10 - engine (Failed)
Errors while running CTest
make: *** [test] Error 8
$ 
```
[LastTest-good.log.txt](https://github.com/gost-engine/engine/files/3535741/LastTest-good.log.txt)

--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -8,7 +8,7 @@ FreeBSD_task:
     PATH: ${PREFIX}/bin:${PATH}
     OPENSSL_BRANCH: master
   script:
-    - pkg install git cmake p5-App-cpanminus gdb
+    - pkg install git cmake p5-App-cpanminus gdb pkgconf
     - sudo cpanm --notest Test2::V0
     - git clone --depth 1 -b ${OPENSSL_BRANCH} https://github.com/openssl/openssl.git
     - cd openssl

--- a/test/run_tests
+++ b/test/run_tests
@@ -1,9 +1,37 @@
 #!/usr/bin/perl
 use TAP::Harness;
 
-if(defined $ENV{'OPENSSL_ROOT_DIR'} && !defined $ENV{'LD_LIBRARY_PATH'}) {
-  $ENV{'LD_LIBRARY_PATH'} = $ENV{'OPENSSL_ROOT_DIR'};
-  $ENV{'PATH'} = "$ENV{'OPENSSL_ROOT_DIR'}/apps:$ENV{'PATH'}";
+if(defined $ENV{'OPENSSL_ROOT_DIR'}) {
+    my $openssl_libdir;
+    my $openssl_bindir;
+
+    if (-d "$ENV{'OPENSSL_ROOT_DIR'}/apps") {
+        # The OpenSSL root dir is an OpenSSL build tree
+        $openssl_bindir = "$ENV{'OPENSSL_ROOT_DIR'}/apps";
+        $openssl_libdir = "$ENV{'OPENSSL_ROOT_DIR'}";
+    } else {
+        # The OpenSSL root dir is an OpenSSL installation tree
+        # Since we're not exactly sure what the library path is (because
+        # multilib), we ask pkg-config
+        local $ENV{PKG_CONFIG_PATH} = "$ENV{'OPENSSL_ROOT_DIR'}/lib/pkgconfig";
+        my $pkgans = `pkg-config --libs-only-L openssl`;
+
+        # If pkg-config failed for any reason, abort.  The tests will most
+        # likely fail anyway because the binary path won't have a matching
+        # library path.
+        die "pkg-config failure: $! (exit code ", $? >> 8, ", signal ", $? & 0xff, ")"
+            if ($? != 0);
+
+        $pkgans =~ s|\R$||;      # Better chomp
+        $pkgans =~ s|^-L||;      # Remove flag from answer
+
+        $openssl_libdir = $pkgans;
+        $openssl_bindir = "$ENV{'OPENSSL_ROOT_DIR'}/bin";
+    }
+    $ENV{'LD_LIBRARY_PATH'} =
+        join(':', $openssl_libdir, split(/:/, $ENV{'LD_LIBRARY_PATH'}));
+    $ENV{'PATH'} =
+        join(':', $openssl_bindir, split(/:/, $ENV{'PATH'}));
 }
 my $harness = TAP::Harness->new();
 exit ($harness->runtests(glob("*.t"))->all_passed() ? 0 : 1);

